.github/workflows/main.yml

```diff
@@ -17,9 +17,69 @@ env:
     target/
 
 jobs:
+  rust_fmt_check:
+    name: Check Rust formatting
+    runs-on: ubuntu-latest
+    steps:
+      - name: Git checkout
+        uses: actions/checkout@v4
+      - name: Setup environment
+        uses: ./.github/actions/setup
+        with:
+          node: ${{ env.NODE_VERSION }}
+      - name: Set environment variables
+        shell: bash
+        run: pnpm zx ./scripts/ci/set-variables.mjs
+      - name: Install Rustfmt
+        uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{ env.RUSTFMT_TOOLCHAIN }}
+          components: rustfmt
+      - name: Check programs
+        run: pnpm programs:format --check
+      - name: Check client
+        run: pnpm clients:rust:format --check
+
+  rust_clippy_check:
+      name: Check Rust clippy
+      runs-on: ubuntu-latest
+      steps:
+        - name: Git checkout
+          uses: actions/checkout@v4
+        - name: Setup environment
+          uses: ./.github/actions/setup
+          with:
+            node: ${{ env.NODE_VERSION }}
+        - name: Set environment variables
+          shell: bash
+          run: pnpm zx ./scripts/ci/set-variables.mjs
+        - name: Install Clippy
+          uses: dtolnay/rust-toolchain@master
+          with:
+            toolchain: ${{ env.CLIPPY_TOOLCHAIN }}
+            components: clippy
+        - name: Check programs
+          run: pnpm programs:clippy
+        - name: Check client
+          run: pnpm clients:rust:clippy
+
+  lint_js:
+    name: Lint JS client
+    runs-on: ubuntu-latest
+    steps:
+      - name: Git checkout
+        uses: actions/checkout@v4
+      - name: Setup environment
+        uses: ./.github/actions/setup
+        with:
+          node: ${{ env.NODE_VERSION }}
+      - name: Lint JS client
+        run: pnpm clients:js:lint
+
   build_programs:
     name: Build programs
     runs-on: ubuntu-latest
+    needs: [rust_fmt_check, rust_clippy_check, lint_js]
```

Comment:
We can keep merge this as is, but keep an eye to see if it makes sense -- clippy can be a bit slow, so we might want to remove this line later

Reply:
although it only took a minute on the most recent run, which should be fine :sweat_smile: